cmd/snap-seccomp/main_test.go

```diff
@@ -0,0 +1,404 @@
+	}
+	defer r.Close()
+
+	for {
+		err = binary.Read(r, nativeEndian(), &rawOp)
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return nil, err
+		}
+		ops = append(ops, rawOp.Disassemble())
+	}
+
+	return ops, nil
+}
+
+func parseBpfInput(s string) (*main.SeccompData /* *seccompData */, error) {
+	// syscall;arch;arg1,arg2...
+	l := strings.Split(s, ";")
+
+	sc, err := seccomp.GetSyscallFromName(l[0])
+	if err != nil {
+		return nil, err
+	}
+
+	var arch uint32
+	if len(l) < 2 || l[1] == "native" {
+		arch = goArchToScmpArch(runtime.GOARCH)
+	} else {
+		arch = goArchToScmpArch(l[1])
+	}
+
+	var syscallArgs [6]uint64
+	if len(l) > 2 {
+		args := strings.Split(l[2], ",")
+		for i := range args {
+			if nr, err := strconv.ParseUint(args[i], 10, 64); err == nil {
+				syscallArgs[i] = nr
+			} else {
+				syscallArgs[i] = main.SeccompResolver[args[i]]
+			}
+		}
+	}
+	sd := &main.SeccompData{}
+	sd.SetArch(arch)
+	sd.SetNr(sc)
+	sd.SetArgs(syscallArgs)
+	return sd, nil
+}
+
+// Endianness detection.
+func nativeEndian() binary.ByteOrder {
+	// Credit matt kane, taken from his gosndfile project.
+	// https://groups.google.com/forum/#!msg/golang-nuts/3GEzwKfRRQw/D1bMbFP-ClAJ
+	// https://github.com/mkb218/gosndfile
+	var i int32 = 0x01020304
+	u := unsafe.Pointer(&i)
+	pb := (*byte)(u)
+	b := *pb
+	isLittleEndian := b == 0x04
+
+	if isLittleEndian {
+		return binary.LittleEndian
+	} else {
+		return binary.BigEndian
+	}
+}
+
+func simulateBpf(c *C, seccompWhitelist, bpfInput string, expected int) {
+	outPath := filepath.Join(c.MkDir(), "bpf")
+	err := main.Compile([]byte(seccompWhitelist), outPath)
+	c.Assert(err, IsNil)
+
+	ops, err := decodeBpfFromFile(outPath)
+	c.Assert(err, IsNil)
+
+	vm, err := bpf.NewVM(ops)
+	c.Assert(err, IsNil)
+
+	bpfSeccompInput, err := parseBpfInput(bpfInput)
+	c.Assert(err, IsNil)
+
+	buf2 := (*[64]byte)(unsafe.Pointer(bpfSeccompInput))
+
+	out, err := vm.Run(buf2[:])
+	c.Assert(err, IsNil)
+	c.Check(out, Equals, expected, Commentf("unexpected result for %q (input %q), got %v expected %v", seccompWhitelist, bpfInput, out, expected))
+}
+
+func (s *snapSeccompSuite) TestCompile(c *C) {
+	// FIXME: this will only work once the following issue is fixed:
+	// https://github.com/golang/go/issues/20556
+	bpf.VmEndianness = nativeEndian()
+
+	for _, t := range []struct {
+		seccompWhitelist string
+		bpfInput         string
+		expected         int
+	}{
+		// special
+		{"@unrestricted", "execve", main.SeccompRetAllow},
+		{"@complain", "execve", main.SeccompRetAllow},
+
+		// trivial alllow
+		{"read", "read", main.SeccompRetAllow},
+		{"read\nwrite\nexecve\n", "write", main.SeccompRetAllow},
+
+		// trivial denial
+		{"read", "execve", main.SeccompRetKill},
+
+		// argument filtering
+		{"read >=2", "read;native;2", main.SeccompRetAllow},
+		{"read >=2", "read;native;1", main.SeccompRetKill},
+
+		{"read <=2", "read;native;2", main.SeccompRetAllow},
+		{"read <=2", "read;native;3", main.SeccompRetKill},
```

Comment:
It sounds like the tests should be written in a clearer manner, I'm not sure (yet) how, I added some comments to hopefully make things a bit clearer. I would love to get your feedback/suggests how to make things easier. I think I'm too deep in that code to see why the test is unclear :/